cloudnew/ec2.go

```diff
@@ -0,0 +1,518 @@
+		return nil, errors.Wrapf(err, "Invalid EC2 settings in distro %s: and %+v", h.Distro.Id, ec2Settings)
+	}
+
+	blockDevices, err := makeBlockDeviceMappings(ec2Settings.MountPoints)
+	if err != nil {
+		return nil, errors.Wrap(err, "error making block device mappings")
+	}
+	h.InstanceType = ec2Settings.InstanceType
+
+	if err := m.client.Create(m.credentials); err != nil {
+		return nil, errors.Wrap(err, "error creating client")
+	}
+
+	var resources []*string
+	if m.onDemand(h) {
+		resources, err = m.spawnOnDemandHost(h, ec2Settings, blockDevices)
+		if err != nil {
+			return nil, errors.Wrap(err, "error spawning on-demand host")
+		}
+	} else if m.spot(h) {
+		resources, err = m.spawnSpotHost(h, ec2Settings, blockDevices)
+		if err != nil {
+			return nil, errors.Wrap(err, "error spawning spot host")
+		}
+	} else {
+		return nil, errors.New("can only spawn on-demand or spot")
+	}
+
+	tags := makeTags(h)
+	tagSlice := []*ec2.Tag{}
+	for tag := range tags {
+		key := tag
+		val := tags[tag]
+		tagSlice = append(tagSlice, &ec2.Tag{Key: &key, Value: &val})
+	}
+	if err = m.client.CreateTags(&ec2.CreateTagsInput{
+		Resources: resources,
+		Tags:      tagSlice,
+	}); err != nil {
+		err = errors.Wrapf(err, "failed to attach tags for %s", h.Id)
+		grip.Error(err)
+		return nil, err
+	}
+
+	grip.Debugf("attached tags for '%s'", h.Id)
+	event.LogHostStarted(h.Id)
+
+	return nil, nil
+}
+
+// CanSpawn indicates if a host can be spawned.
+func (m *ec2Manager) CanSpawn() (bool, error) {
+	return true, nil
+}
+
+// GetInstanceStatus returns the current status of an EC2 instance.
+func (m *ec2Manager) GetInstanceStatus(h *host.Host) (cloud.CloudStatus, error) {
+	if err := m.client.Create(m.credentials); err != nil {
+		return cloud.StatusUnknown, errors.Wrap(err, "error creating client")
+	}
+	if m.onDemand(h) {
+		info, err := m.getInstanceInfo(h.Id)
+		if err != nil {
+			return cloud.StatusUnknown, err
+		}
+		return ec2StatusToEvergreenStatus(*info.State.Name), nil
+	} else if m.spot(h) {
+		return m.getSpotInstanceStatus(h.Id)
+	}
+	return cloud.StatusUnknown, errors.New("type must be on-demand or spot")
+}
+
+// TerminateInstance terminates the EC2 instance.
+func (m *ec2Manager) TerminateInstance(h *host.Host) error {
+	// terminate the instance
+	if h.Status == evergreen.HostTerminated {
+		err := errors.Errorf("Can not terminate %s - already marked as "+
+			"terminated!", h.Id)
+		grip.Error(err)
+		return err
+	}
+
+	if err := m.client.Create(m.credentials); err != nil {
+		return errors.Wrap(err, "error creating client")
+	}
+
+	if m.spot(h) {
+		canTerminate, err := m.cancelSpotRequest(h.Id)
+		if err != nil {
+			return errors.Wrap(err, "error canceling spot request")
+		}
+		// the spot request wasn't fulfilled, so don't attempt to terminate
+		if !canTerminate {
+			return nil
+		}
+	}
+
+	resp, err := m.client.TerminateInstances(&ec2.TerminateInstancesInput{
+		InstanceIds: []*string{&h.Id},
+	})
+	if err != nil {
+		return err
```

Comment:
Does this handle if the instance was terminated externally? I believe the API will return an instance not found error here

Reply:
Fixed.

Reply:
What was the change?

Reply:
https://github.com/bsamek/evergreen/blob/4b867ff6901db6d7a9da43858adb9affaa244047/cloud/new_ec2.go#L370-L374